haystack/core/serialization.py

```diff
@@ -2,11 +2,32 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 import inspect
-from typing import Any, Dict, Type
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Optional, Type
```

Comment:
We need to import from `from collections.abc import Callable` as importing from `typing` is deprecated as [PEP 585](https://peps.python.org/pep-0585/).

```suggestion
from typing import Any, Dict, Optional, Type
```

This will make the below type definition fail too.